xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp

```diff
@@ -1598,7 +1598,24 @@ int dvd_inputstreamnavigator_cb_seek(void * p_stream, uint64_t i_pos)
 int dvd_inputstreamnavigator_cb_read(void * p_stream, void * buffer, int i_read)
 {
   CDVDInputStreamFile *lpstream = reinterpret_cast<CDVDInputStreamFile*>(p_stream);
-  return lpstream->Read(reinterpret_cast<uint8_t *>(buffer), i_read);
+
+  int i_ret = 0;
+  while (i_ret < i_read)
+  {
+    int i_r;
+    i_r = lpstream->Read(reinterpret_cast<uint8_t *>(buffer) + i_ret, i_read - i_ret);
+    if (i_r < 0)
+    {
+      CLog::Log(LOGERROR,"read error dvd_inputstreamnavigator_cb_read");
+      return i_r;
+    }
+    if (i_r == 0)
```

Comment:
@ace20022 - forgot to mention that you will not always read "i_read" bytes. Sometimes at the end of a file you will read an incomplete block and i_r will be 0 before you read i_read bytes... The problem with FTP streams was apparently that one read operation only read e.g. 1600 bytes instead of 2048 and you needed to try again as long as you could read more.